File: 1term/10.22/4/4/stack.cpp
#include <stdio.h>
#include "stack.h"

struct Stack
{
	ListElement *head;
};

struct ListElement
{
	ElementType value;
	ListElement *next;
};

Stack *createStack()
{
	Stack *newStack = new Stack;
	newStack->head = NULL;
	return newStack;
}

void addElementStack(Stack *stack, ElementType value)
{
	ListElement *newListElement = new ListElement;
	newListElement->value = value;
	newListElement->next = stack->head;
	stack->head = newListElement;
}

ElementType valueHeadStack(Stack *stack)
{
	return stack->head->value;
}

void deleteStackElement(Stack *stack)
{
	if (stack == NULL || stack->head == NULL)
		return;
	ListElement *temp = stack->head;
	stack->head = stack->head->next;
	delete temp;
}

void deleteStack(Stack *stack)
{
	while (stack->head != NULL)
		deleteStackElement(stack);
	delete stack;
}

bool isEmptyStack(Stack *stack)
{
	return (stack->head == NULL);
}

File: 2term/02.19/1stack/1stack/stack.cs
using System;

using ElementType = System.Int32;

namespace NStack
{
    public class Stack
    {
        public Stack()
        {
            head = null;
        }

        // add element newValue to head.
        public void Push(ElementType newValue)
        {
            ListElement temp = new ListElement(newValue, head);
            head = temp;
        }

        // delete element in head and return value.
        public ElementType Pop()
        {
            if (head == null)
                return -1;
            ElementType val = head.Value();
            head = head.Next();
            return val; 
        }

        // print stack.
        public void Print()
        {
            if (head == null)
            {
                Console.WriteLine("Stack is empty");
                return;
            }
            ListElement pos = head;
            for (pos = head; pos.Next() != null; pos = pos.Next())
                Console.Write("{0} -> ", pos.Value());
            Console.Write("{0}", pos.Value()); 
            Console.WriteLine();
        }

       private ListElement head;

       private class ListElement
       {
           public ListElement()
           { }

           public ListElement(ElementType value, ListElement next)
           {
               this.value = value;
               this.next = next;
           }

           public ElementType Value()
           {
               return this.value;
           }

           public ListElement Next()
           {
               return this.next;
           }

           private ElementType value;
           private ListElement next;
       }

    }
}
